File: examples/25_mcp_single_tool_gateway.py
"""Example 25: binding ONE gateway-federated MCP tool — the singular ``mcp_tool_factory``.

Example 23 shows the plural ``mcp_tool_factories(...)``: connect at build time,
DISCOVER every tool the servers expose, slice the returned dict per node. This
example shows its counterpart for when you already KNOW the tool's name and want
the build path to stay OFFLINE:

    mcp_tool_factories(servers)          # plural  — asks the server (discovery,
                                         #           connects at build, fail-fast)
    mcp_tool_factory(key, spec,          # singular — tells the server (declared
                     tool_name=...)      #           name, ZERO network at build)

The mnemonic: the plural asks the server, the singular tells the server.

The scenario is the gateway-federated case that motivated the helper (filed by a
real consumer). An MCP gateway (e.g. IBM ContextForge) re-exposes a peer's tool
NAMESPACED as ``<peer>-<tool>`` — here the demo server exposes
``crm-perplexity_research``. The node, written long before the gateway config,
binds the bare name ``Tool("perplexity_research")``. The singular factory bridges
the two: it binds the declared namespaced name and renames the fetched tool back
to the bare binding, and nothing connects until the agent's first tool call.

Three beats:

  1. OFFLINE BUILD — constructing the factory (and compiling the graph) performs
     zero network I/O. We prove it by building a factory against a spec that
     CANNOT connect (a nonexistent command): construction still succeeds, because
     the connect lives inside the returned async factory, deferred to first await.
     This is what keeps a consumer's compile()-time and test paths offline.

  2. NAMESPACED -> BARE RENAME — ``tool_name="crm-perplexity_research",
     rename_to="perplexity_research"`` makes both the factory key and the
     LLM-facing ``tool.name`` match the node's fixed ``Tool`` spec.

  3. PER-RUN IDENTITY, SAME PATH — the singular takes the same ``token_provider``
     as the plural; over stdio the token rides as a tool argument the server
     echoes back (``acting_as``), exactly as in example 23 beat 3.

The MCP layer is real (the same stdio demo server as examples 23/24); the only
fake is the LLM, so the example is deterministic and keyless.

Run (needs the mcp-examples extra; no API keys, no network beyond the local
subprocess):
    uv run --extra dev --extra mcp-examples python examples/25_mcp_single_tool_gateway.py
"""

from __future__ import annotations

import asyncio
import sys
from pathlib import Path
from typing import Any

from langchain_core.messages import AIMessage, ToolMessage
from pydantic import BaseModel

from neograph import Tool, ToolInteraction, arun, compile, construct_from_functions, node
from neograph_mcp import StdioServer, mcp_tool_factory

DEMO_SERVER = str((Path(__file__).parent / "_mcp_demo_server.py").resolve())

CRM_GATEWAY = StdioServer(command=sys.executable, args=[DEMO_SERVER])


# ── Schemas ──────────────────────────────────────────────────────────────────


class ResearchNote(BaseModel, frozen=True):
    query: str
    finding: str


class PerplexityResult(BaseModel, frozen=True):
    """Client-side model for the gateway tool's structuredContent — declaring it as
    output_model= makes typed_result the model, not a raw content-block list."""

    query: str
    acting_as: str  # the per-run identity the server echoed


# ── Fake LLM (the ONLY fake — the MCP layer is real) ─────────────────────────
# One agent turn: call perplexity_research, then finalize from its result.


class _ResearchFake:
    _NOTE = ResearchNote(query="acme", finding="Acme renewal research complete.")

    def __init__(self) -> None:
        self._structured = False

    def bind_tools(self, tools: list) -> _ResearchFake:
        return self

    def abind_tools(self, *a: Any, **k: Any) -> _ResearchFake:
        return self

    def with_structured_output(self, model: type[BaseModel], **k: Any) -> _ResearchFake:
        clone = _ResearchFake()
        clone._structured = True
        return clone

    def invoke(self, messages: list, **k: Any) -> Any:
        if self._structured:
            return self._NOTE
        if not any(isinstance(m, ToolMessage) for m in messages):
            msg = AIMessage(content="")
            # The model calls the BARE name — it never sees the gateway namespace.
            msg.tool_calls = [{"name": "perplexity_research", "args": {"query": "acme"}, "id": "p1"}]
            return msg
        return AIMessage(content=self._NOTE.model_dump_json())

    async def ainvoke(self, *a: Any, **k: Any) -> Any:
        return self.invoke(*a, **k)


def llm_factory(tier: str) -> Any:
    return _ResearchFake()


def prompt_compiler(template: str, data: Any, **kw: Any) -> list[dict[str, str]]:
    return [{"role": "user", "content": template}]


# ── The node: written against the BARE tool name ─────────────────────────────


@node(
    mode="agent",
    outputs={"result": ResearchNote, "tool_log": list[ToolInteraction]},
    model="research",
    prompt="research/note",
    tools=[Tool("perplexity_research", budget=1, idempotent=True)],
)
def research() -> ResearchNote:  # body unused — the LLM drives via prompt= + tools=
    ...


pipeline = construct_from_functions("gateway-research", [research])


# ── Demos ─────────────────────────────────────────────────────────────────────


def demo_offline_build() -> None:
    """Beat 1: construction connects NOWHERE — even an unreachable spec builds."""
    print("=" * 66)
    print("BEAT 1: offline build — zero network at construction")
    print("=" * 66)

    unreachable = StdioServer(command="/nonexistent/not-a-real-gateway", args=["--nope"])
    factory = mcp_tool_factory(
        "crm",
        unreachable,
        tool_name="crm-perplexity_research",
        rename_to="perplexity_research",
    )
    # No subprocess spawned, no get_tools called. The connect lives inside the
    # returned async factory and fires only on its first await.
    assert asyncio.iscoroutinefunction(factory)
    print("\nBuilt a factory against an UNREACHABLE spec — no error, no connect.")
    print("Discovery is a connect, so the plural mcp_tool_factories() would have")
    print("failed here at build time. The singular defers: your compile() and")
    print("test paths stay offline; a wrong name fails loud at first tool call.")


async def demo_rename_and_identity() -> None:
    """Beats 2 + 3: namespaced->bare rename and per-run identity, end to end."""
    print("\n" + "=" * 66)
    print("BEAT 2/3: gateway rename (crm-perplexity_research -> perplexity_research)")
    print("=" * 66)

    factory = mcp_tool_factory(
        "crm",
        CRM_GATEWAY,
        tool_name="crm-perplexity_research",  # the name the GATEWAY exposes
        rename_to="perplexity_research",  # the name the NODE binds
        token_provider=lambda configurable: configurable.get("mcp_auth", "anon"),
        # TYPED RESULT: rehydrate the server's structuredContent into our model, so
        # typed_result IS a PerplexityResult — no hand-parsing of content blocks.
        output_model=PerplexityResult,
    )

    # Still offline: compiling binds the factory under the bare name; nothing has
    # connected yet.
    graph = compile(
        pipeline,
        llm_factory=llm_factory,
        prompt_compiler=prompt_compiler,
        tool_factories={"perplexity_research": factory},
    )
    print("\nCompiled the graph — still zero MCP connects.")

    # The first tool call inside the run is where the connect finally happens.
    result = await arun(
        graph,
        input={"node_id": "GW-1"},
        config={"configurable": {"mcp_auth": "operator-A"}},
    )
    note: ResearchNote = result["research_result"]
    tool_log: list[ToolInteraction] = result["research_tool_log"]

    call = tool_log[0]
    payload = call.typed_result  # a PerplexityResult model, not raw content blocks
    print(f"\nLLM-facing tool name : {call.tool_name}")
    print(f"server-side identity : acting_as={payload.acting_as!r}")
    print(f"final result         : {note.finding}")
    assert call.tool_name == "perplexity_research"  # the bare binding, not <peer>-<tool>
    assert payload.acting_as == "operator-A"  # same token path as the plural
    assert payload.query == "acme"


async def main() -> None:
    demo_offline_build()
    await demo_rename_and_identity()
    print("\n" + "=" * 66)
    print("Done. One declared tool, bound offline, renamed to the node's bare")
    print("Tool(name), carrying per-run identity — no discovery connect anywhere.")
    print("=" * 66)


if __name__ == "__main__":
    asyncio.run(main())

File: docs/design/ts-parity/25_mcp_single_tool_gateway.ts
// HYPOTHETICAL TypeScript port of examples/25_mcp_single_tool_gateway.py
// Grounds a feature-parity analysis against docs/design/typescript-port.md (AD-0 transformer form).
// This is NOT meant to compile or run — it is what the TS DX WOULD look like, with
// '// PARITY:' notes wherever the TS diverges or the proposed API has no answer.
//
// Python source teaches: the SINGULAR neograph_mcp.mcp_tool_factory — bind ONE
// gateway-federated MCP tool with an OFFLINE build (zero network at compile()),
// namespaced->bare rename, and per-run identity via token_provider.

import { z } from "zod";
import { AIMessage, ToolMessage } from "@langchain/core/messages";
import { node, Tool, ToolInteraction, compile, run, constructFromFunctions } from "@neograph/core";

// PARITY(API GAP — headline): the ENTIRE neograph_mcp package is absent from the
// typescript-port.md Feature Parity Matrix. StdioServer / HttpServer /
// mcp_tool_factory (singular) / mcp_tool_factories (plural) / token_provider /
// output_model rehydration / gateway rename / mcp_resource_fetcher /
// mcp_run_context have NO proposed TS design. The imports below are invented
// (`@neograph/mcp`). @langchain/mcp-adapters exists in JS, so a port is BUILDABLE,
// but the whole abstraction layer this example demonstrates is undesigned.
import { StdioServer, mcpToolFactory, type ToolFactory } from "@neograph/mcp";

const DEMO_SERVER = new URL("./_mcp_demo_server.py", import.meta.url).pathname;
const CRM_GATEWAY = new StdioServer({ command: process.execPath, args: [DEMO_SERVER] });

// ── Schemas ──────────────────────────────────────────────────────────────────
// PARITY(Direct-ish): Python's `class X(BaseModel, frozen=True)` -> Zod (AD-1).
// `frozen=True` immutability has no runtime Zod analog; use `.readonly()` or rely
// on `as const`. The nominal class identity Python uses for producer-type matching
// becomes structural Zod-schema comparison (AD-3) — a known Redesign for the validator.
const ResearchNote = z.object({
  query: z.string(),
  finding: z.string(),
});
type ResearchNote = z.infer<typeof ResearchNote>;

// Client-side model for the gateway tool's structuredContent. Declaring it as
// output_model= makes typed_result THIS model, not a raw content-block list.
const PerplexityResult = z.object({
  query: z.string(),
  acting_as: z.string(), // the per-run identity the server echoed
});
type PerplexityResult = z.infer<typeof PerplexityResult>;

// ── Fake LLM (the ONLY fake — the MCP layer is real) ─────────────────────────
// One agent turn: call perplexity_research, then finalize from its result.
//
// PARITY(Redesign): Python duck-types the LLM (any object with bind_tools/
// with_structured_output/invoke/ainvoke). LangChain.js has no structural-typing
// escape hatch — a fake must satisfy the Runnable/BaseChatModel interface (or be
// cast `as unknown as BaseChatModel`). `with_structured_output(Model)` -> Zod
// `withStructuredOutput(schema)`. Python returns a Pydantic instance straight from
// `invoke`; JS returns a plain parsed object.
class ResearchFake {
  private static NOTE: ResearchNote = { query: "acme", finding: "Acme renewal research complete." };
  private structured = false;

  bindTools(_tools: unknown[]): ResearchFake {
    return this;
  }

  withStructuredOutput(_schema: z.ZodTypeAny): ResearchFake {
    const clone = new ResearchFake();
    clone.structured = true;
    return clone;
  }

  // PARITY(Simplification): JS is async-native, so Python's invoke/ainvoke split
  // collapses to ONE `invoke`. See the arun/run note below — a genuine TS win.
  async invoke(messages: unknown[]): Promise<unknown> {
    if (this.structured) return ResearchFake.NOTE;
    if (!messages.some((m) => m instanceof ToolMessage)) {
      // The model calls the BARE name — it never sees the gateway namespace.
      return new AIMessage({
        content: "",
        tool_calls: [{ name: "perplexity_research", args: { query: "acme" }, id: "p1" }],
      });
    }
    return new AIMessage({ content: JSON.stringify(ResearchFake.NOTE) });
  }
}

function llmFactory(_tier: string): ResearchFake {
  return new ResearchFake();
}

function promptCompiler(template: string, _data: unknown): Array<{ role: string; content: string }> {
  return [{ role: "user", content: template }];
}

// ── The node: written against the BARE tool name ─────────────────────────────
//
// PARITY(Redesign, but AD-0 buys NOTHING here): the doc's centerpiece — the
// signature-extraction transformer that preserves "signature IS the DAG" — is
// IRRELEVANT to this example. `research()` takes ZERO parameters (an agent leaf)
// and its body is DEAD (`...`), so there is no signature to extract. Everything
// load-bearing (mode, dict-form outputs, tools) is config-object data that ports
// identically WITH OR WITHOUT the transformer. This example is orthogonal to AD-0.
//
// PARITY(API GAP): dict-form `outputs={"result": X, "tool_log": list[ToolInteraction]}`
// with a framework-COLLECTED secondary key (tool_log) is not in the matrix. The
// demand-driven tool_log collection ("no consumer -> no overhead") must be
// re-specified. `list[ToolInteraction]` -> `z.array(ToolInteraction.schema)`.
const research = node(
  {
    mode: "agent",
    outputs: { result: ResearchNote, tool_log: z.array(ToolInteraction.schema) },
    model: "research",
    prompt: "research/note",
    tools: [new Tool({ name: "perplexity_research", budget: 1, idempotent: true })],
  },
  // PARITY: dead body must still be written as a callback; there is no `...`
  // stub form and no AST dead-body warning (matrix marks that Skip / ESLint-later).
  (): ResearchNote => {
    throw new Error("unreachable — the LLM drives via prompt= + tools=");
  },
);

// PARITY(Direct, ~2 days per matrix): construct_from_functions -> constructFromFunctions.
const pipeline = constructFromFunctions("gateway-research", [research]);

// ── Demos ─────────────────────────────────────────────────────────────────────

// Beat 1: construction connects NOWHERE — even an unreachable spec builds.
function demoOfflineBuild(): void {
  const unreachable = new StdioServer({ command: "/nonexistent/not-a-real-gateway", args: ["--nope"] });
  const factory: ToolFactory = mcpToolFactory("crm", unreachable, {
    toolName: "crm-perplexity_research",
    renameTo: "perplexity_research",
  });

  // PARITY(Redesign / no clean idiom — load-bearing for THIS example): Python
  // PROVES the offline-build guarantee with `asyncio.iscoroutinefunction(factory)`
  // — a first-class runtime predicate that the connect is deferred inside the
  // returned coroutine fn. JS has no honest equivalent: any function may return a
  // Promise, and `factory.constructor.name === "AsyncFunction"` is fragile (breaks
  // the moment the factory is wrapped/bound/transpiled to a generator by a
  // downlevel target). The teaching assertion of Beat 1 has no direct TS port.
  console.assert(factory.constructor.name === "AsyncFunction"); // fragile; see note

  console.log("Built a factory against an UNREACHABLE spec — no error, no connect.");
}

// Beats 2 + 3: namespaced->bare rename and per-run identity, end to end.
async function demoRenameAndIdentity(): Promise<void> {
  const factory: ToolFactory = mcpToolFactory("crm", CRM_GATEWAY, {
    toolName: "crm-perplexity_research", // the name the GATEWAY exposes
    renameTo: "perplexity_research", // the name the NODE binds
    // PARITY(Direct): a closure reading per-run config. Python's
    // `lambda configurable: configurable.get("mcp_auth", "anon")` ->
    // `(cfg) => cfg.mcp_auth ?? "anon"`. Clean.
    tokenProvider: (cfg: Record<string, unknown>) => (cfg.mcp_auth as string) ?? "anon",
    // PARITY(API GAP): output_model rehydration of the server's structuredContent
    // into a typed model. No proposed design. With Zod it would be a `.parse()`
    // at the tool boundary, but the whole seam (rehydrate MCP content-block JSON ->
    // typed instance so typed_result IS PerplexityResult) is unspecified.
    outputModel: PerplexityResult,
  });

  // PARITY(API GAP): the `tool_factories=` compile channel takes an ASYNC factory
  // with a `(config, tool_config) => Tool` signature (per-call identity + per-call
  // idempotency channel). The matrix lists "Registries (scripted, condition, tool)"
  // as Direct but says nothing about this deferred-async-factory contract.
  const graph = compile(pipeline, {
    llmFactory,
    promptCompiler,
    toolFactories: { perplexity_research: factory },
  });

  // PARITY(Simplification — TS WIN): Python splits run/arun; JS is async-native so
  // there is ONE `run` (awaitable). The matrix's Runner row doesn't note this, but
  // the arun/run duplication simply disappears in TS.
  const result = await run(graph, {
    input: { node_id: "GW-1" },
    config: { configurable: { mcp_auth: "operator-A" } },
  });

  // PARITY(Redesign): dict result keys are stringly-typed. Python reads
  // result["research_result"]; TS `result["research_result"]` is untyped unless
  // compile() threads an output type param. `{node}_{key}` state-field naming
  // must be reproduced.
  const note = result["research_result"] as ResearchNote;
  const toolLog = result["research_tool_log"] as ToolInteraction[];

  const call = toolLog[0];
  // PARITY(Redesign): Python `ToolInteraction.typed_result: Any`. In TS that field
  // is `unknown` (the "no-Any" ban has no TS analog beyond `unknown`), so the
  // caller MUST narrow/cast — there is no typed retrieval path in the proposed API.
  const payload = call.typed_result as PerplexityResult;

  console.log(`LLM-facing tool name : ${call.tool_name}`);
  console.log(`server-side identity : acting_as=${payload.acting_as}`);
  console.log(`final result         : ${note.finding}`);
  console.assert(call.tool_name === "perplexity_research"); // bare binding, not <peer>-<tool>
  console.assert(payload.acting_as === "operator-A"); // same token path as the plural
  console.assert(payload.query === "acme");
}

async function main(): Promise<void> {
  demoOfflineBuild();
  await demoRenameAndIdentity();
}

void main();
